packages/SwingSet/docs/vat-environment.md

```diff
@@ -0,0 +1,113 @@
+* `URL`, `URLSearchParams` are not available
+* `WebAssembly` is not available, as neat as that would be
+* `TextEncoder` and `TextDecoder` are not available
+
+Some names look like globals, but are really part of the tools that define modules: imports, exports, and metadata. Modules start as files on disk, but are then bundled together into an archive before being loaded into a vat. Several standard functions are used by the bundling tool to locate all the other modules that must be included. These are not a part of SES, but are allowed in module source code, and are translated or removed before execution finally happens:
+
+* `import` and `export` syntax are allowed in ESM-style modules (which is preferred over CommonJS). These are not globals per se, but rather top-level syntax which defines the module graph.
+* `require`, `module`, `module.exports`, and `exports` are allowed in CommonJS-style modules, and should work as expected, however new code should be written as ESM modules. They are either consumed by the the bundling process, provided (in some form) by the execution environment, or otherwise rewritten to work sensibly
+* `__dirname` and `__filename` are not provided
+* The dynamic import expression (`await import('name')`) is currently prohibited in vat code, but a future SES implementation may allow it.
+
+Node.js has a [large collection](https://nodejs.org/dist/latest-v14.x/docs/api/) of "built-in modules", like `http` and `crypto`. Some are clearly platform-specific (like `v8`), while others are not so obvious (`stream`). All of these are accessed by importing a module (`const v8 = require('v8')` in CommonJS modules, or `import v8 from 'v8'` in ESM modules). These modules are built out of native code (C++), not plain JS.
+
+None of these built-in modules are available to Vat code. `require` or `import` can be used on modules that contain pure JS, but not on modules that include native code. To allow a vat to exercise authority which comes from a built-in module, you will have to write a "device", give that device an endowment with the built-in module's functions, then have the vat send messages to the device.
+
+Browser environments also have a huge list of [other features](https://developer.mozilla.org/en-US/docs/Web/API) which are presented as names in the global scope (some of which were also added to Node.js). None of these are available in a SES environment. The most surprising removals include `atob`, `TextEncoder`, and `URL`.
+
+`debugger` is a first-class JavaScript statement, and behaves as expected in vat code.
+
+## Other Changes
+
+The [shim](https://github.com/Agoric/SES-shim/) which provides our SES environment is not as fully-featured as a native implemenation could be. As a result, some forms of code cannot be used yet. These restrictions should be lifted once the JS engine you use can provide SES natively:
```

Comment:
"which provides" -> "that provides"

I'd drop the "could be".

Do you mean to have a : at the end? Since there's no list, but rather a whole new section, this should either be a period is you didn't intent for a list-like thing, or rephrased to "natively. The following sections describe these still restricted code elements." (or similar if I didn't get how to refer to them correct).

Reply:
Yeah, the following sections are a list (I'm relying upon Markdown to display them with sub-headings or something). What about "These restrictions, described in the following sections, should be lifted once ... natively." ?

Reply:
Looks good.